DS-15222: Add LogpushJob_Full test to test all fields

soheiokamoto opened the pull request:
This adds `LogpushJob_Full` test to test all fields.

This is to demonstrate what changes need to be made on `model.go` and `schema.go`, which are manually done here to make the t

Notable changes are the following:
 1. Remove `Computed` and `Default` from some fields to match with Logpush API.
 2. Added `Computed` for `Enabled` field, as Logpush API always return value (default `false`).
 2. Change `max_upload_*` to be integer, with range validator.
 3. Adding `no_refresh` attribute for the fields that cause "drift" due to Logpush API json `omitempty`.

Some changes would need to be applied to Logpush API OpenAPI schema to make same changes on the next SDK generation.

soheiokamoto commented:
Added a commit to tests more.

vaishakdinesh commented on internal/services/logpush_job/schema.go:
```diff
@@ -132,25 +124,38 @@ func ResourceSchema(ctx context.Context) schema.Schema {
 				Optional:    true,
-				Validators: []validator.Float64{
-					float64validator.OneOf(0),
+				Validators: []validator.Int64{
```
I think these changes can be made in the openapi schema. We don't to add custom code unless its absolutely necessary.

soheiokamoto replied:
Yes, that makes sense.
I was just trying to figure out what changes would be needed here first, so that I can work on OpenAPI or SDK generation to match the changes where applicable.

soheiokamoto replied:
As far as I can tell, the source OpenAPI schema seems correct , so I'm not quite sure how to fix this from OpenAPI side:
```
      "logpush_max_upload_bytes": {
        "description": "The maximum uncompressed file size of a batch of logs. This setting value must be between `5 MB` and `1 GB`, or `0` to disable it. Note that you cannot set a minimum file size; this means that log files may be much smaller than this batch size.",
        "example": 5000000,
        "nullable": true,
        "type": "integer",
        "x-auditable": true,
        "anyOf": [
          { "enum": [0], "type": "number" },
          { "maximum": 1000000000, "minimum": 5000000 }
        ]
      },
      "logpush_max_upload_interval_seconds": {
        "default": 30,
        "description": "The maximum interval in seconds for log batches. This setting must be between 30 and 300 seconds (5 minutes), or `0` to disable it. Note that you cannot specify a minimum interval for log batches; this means that log files may be sent in shorter intervals than this.",
        "example": 30,
        "nullable": true,
        "type": "integer",
        "x-auditable": true,
        "anyOf": [
          { "enum": [0], "type": "number" },
          { "maximum": 300, "minimum": 30 }
        ]
      },
      "logpush_max_upload_records": {
        "default": 100000,
        "description": "The maximum number of log lines per batch. This setting must be between 1000 and 1,000,000 lines, or `0` to disable it. Note that you cannot specify a minimum number of log lines per batch; this means that log files may contain many fewer lines than this.",
        "example": 1000,
        "nullable": true,
        "type": "integer",
        "x-auditable": true,
        "anyOf": [
          { "enum": [0], "type": "number" },
          { "maximum": 1000000, "minimum": 1000 }
        ]
      },
```

soheiokamoto replied:
Found the commit when this was changed, actually: https://github.com/cloudflare/terraform-provider-cloudflare/commit/39676a0de28e699286354db0cae59ba083860ebc

I don't know why it did that, trying to find out.

soheiokamoto replied:
Ah. These got changed for some reason when we added enum 0.

Before, it was below, and was fine (`Int64`):
```
      "logpush_max_upload_bytes": {
        "description": "The maximum uncompressed file size of a batch of logs. This setting value must be between `5 MB` and `1 GB`, or `0` to disable it. Note that you cannot set a minimum file size; this means that log files may be much smaller than this batch size. This parameter is not available for jobs with `edge` as its kind.",
        "example": 5000000,
        "maximum": 1000000000,
        "minimum": 5000000,
        "nullable": true,
        "type": "integer",
        "x-auditable": true
      },
      "logpush_max_upload_interval_seconds": {
        "default": 30,
        "description": "The maximum interval in seconds for log batches. This setting must be between 30 and 300 seconds (5 minutes), or `0` to disable it. Note that you cannot specify a minimum interval for log batches; this means that log files may be sent in shorter intervals than this. This parameter is only used for jobs with `edge` as its kind.",
        "example": 30,
        "maximum": 300,
        "minimum": 30,
        "nullable": true,
        "type": "integer",
        "x-auditable": true
      },
      "logpush_max_upload_records": {
        "default": 100000,
        "description": "The maximum number of log lines per batch. This setting must be between 1000 and 1,000,000 lines, or `0` to disable it. Note that you cannot specify a minimum number of log lines per batch; this means that log files may contain many fewer lines than this. This parameter is not available for jobs with `edge` as its kind.",
        "example": 1000,
        "maximum": 1000000,
        "minimum": 1000,
        "nullable": true,
        "type": "integer",
        "x-auditable": true
      },
```

vaishakdinesh commented:
The test and `no_refresh` lgtm. 
But the model and schema changes can be made in the openapi spec.

soheiokamoto commented:
> The test and `no_refresh` lgtm. But the model and schema changes can be made in the openapi spec.

Thanks for the review.

Going to work on OpenAPI side now.

soheiokamoto commented:
OpenAPI schema for Logpush job was updated.

The changes are reflected in next branch: https://github.com/cloudflare/terraform-provider-cloudflare/commit/3120f94b6af470eee1c50468cadcc5786268e381

Rebased and squashed commits for this PR.

Tests are failing locally feemingly unrelated to this PR (and preventing to run actual tests:
```
soheiokamoto/terraform-provider-cloudflare [sohei/DS-15222-add-logpush-job-tests-full] $ go test ./internal/services/logpush_job -run "^TestAccCloudflareLogpushJob_Basic$" -v -count 1

# github.com/cloudflare/terraform-provider-cloudflare/internal/services/email_security_block_sender
internal/services/email_security_block_sender/data_source_model.go:52:10: params.Pattern undefined (type email_security.SettingBlockSenderListParams has no field or method Pattern)
internal/services/email_security_block_sender/list_data_source_model.go:43:10: params.Pattern undefined (type email_security.SettingBlockSenderListParams has no field or method Pattern)
# github.com/cloudflare/terraform-provider-cloudflare/internal/services/email_security_trusted_domains
internal/services/email_security_trusted_domains/data_source_model.go:59:10: params.Pattern undefined (type email_security.SettingTrustedDomainListParams has no field or method Pattern)
internal/services/email_security_trusted_domains/list_data_source_model.go:50:10: params.Pattern undefined (type email_security.SettingTrustedDomainListParams has no field or method Pattern)
# github.com/cloudflare/terraform-provider-cloudflare/internal/services/workers_deployment
internal/services/workers_deployment/data_source.go:71:3: cannot use data.DeploymentID.ValueString() (value of type string) as workers.ScriptDeploymentGetParams value in argument to d.client.Workers.Scripts.Deployments.Get
internal/services/workers_deployment/data_source.go:72:3: cannot use params (variable of type workers.ScriptDeploymentGetParams) as option.RequestOption value in argument to d.client.Workers.Scripts.Deployments.Get: workers.ScriptDeploymentGetParams does not implement requestconfig.RequestOption (missing method Apply)
internal/services/workers_deployment/resource.go:117:3: cannot use data.ID.ValueString() (value of type string) as workers.ScriptDeploymentGetParams value in argument to r.client.Workers.Scripts.Deployments.Get
internal/services/workers_deployment/resource.go:118:3: cannot use workers.ScriptDeploymentGetParams{…} (value of type workers.ScriptDeploymentGetParams) as option.RequestOption value in argument to r.client.Workers.Scripts.Deployments.Get: workers.ScriptDeploymentGetParams does not implement requestconfig.RequestOption (missing method Apply)
internal/services/workers_deployment/resource.go:153:49: r.client.Workers.Scripts.Deployments.Delete undefined (type *workers.ScriptDeploymentService has no field or method Delete)
internal/services/workers_deployment/resource.go:157:11: undefined: workers.ScriptDeploymentDeleteParams
internal/services/workers_deployment/resource.go:197:3: cannot use path_deployment_id (variable of type string) as workers.ScriptDeploymentGetParams value in argument to r.client.Workers.Scripts.Deployments.Get
internal/services/workers_deployment/resource.go:198:3: cannot use workers.ScriptDeploymentGetParams{…} (value of type workers.ScriptDeploymentGetParams) as option.RequestOption value in argument to r.client.Workers.Scripts.Deployments.Get: workers.ScriptDeploymentGetParams does not implement requestconfig.RequestOption (missing method Apply)
# github.com/cloudflare/terraform-provider-cloudflare/internal/services/image
internal/services/image/list_data_source_model.go:34:10: params.Creator undefined (type images.V1ListParams has no field or method Creator)
FAIL    github.com/cloudflare/terraform-provider-cloudflare/internal/services/logpush_job [build failed]
FAIL
```

soheiokamoto commented:
The linting and tests have been failing, becuase `terraform-provider-cloudflare` `next` branch is still based on `cloudflare-go/v4` version 4.6.0 (latest `main` there), but some changes are dependent on 4.7.0 (in the `next` branch there), which is not merged yet: https://github.com/cloudflare/cloudflare-go/pull/4173

Tests pass locally now with the latest commit on the vendor package.

Temporarily added commit to update the vendor with the latest commit for 4.7.0 (since the tag is not ready yet): https://github.com/cloudflare/terraform-provider-cloudflare/pull/5771/commits/061d5b4a8ec30602468cdf71e2501ed0189e63ad

The commit should be removed as soon as the `next` branch upgrades to 4.7.0, but wanted to have CI tests pass here.

soheiokamoto commented on go.mod:
```diff
@@ -10,7 +10,7 @@ require (
 	github.com/aws/aws-sdk-go-v2/service/s3 v1.63.0
 	github.com/cloudflare/cloudflare-go v0.104.0
-	github.com/cloudflare/cloudflare-go/v4 v4.6.0
+	github.com/cloudflare/cloudflare-go/v4 v4.6.1-0.20250723225234-a89c233b7596
```
Removed the commit.

soheiokamoto commented:
Removed the temporary commits that were added in attempt to make CI tests pass.

musa-cf commented:
Thank you for adding these tests, @soheiokamoto! LGTM
